src/ai/defaultai.cc

```diff
@@ -294,9 +289,12 @@ void DefaultAI::think() {
 		}
 	}
 
-	// Here we decide how many jobs will be run now (none - 5)
-	// in case no job is due now, it can be zero
-	uint32_t jobs_to_run_count = (delay_time < 0) ? 0 : 1;
+	// Here we decide limit for jobs will be run now
+	// generelly 3, if situation is wrong: 4
+	uint32_t jobs_to_run_count = std::min<uint32_t>((delay_time > 3000) ? 5 : 3, taskPool.size());
+
+	// we need to get more urgent jobs to the front
+	sort_task_pool(jobs_to_run_count);
```

Comment:
this would run sort taskpool more often then necessary. In current design we have only 1 sorting cycle as normally only 1 job is run per each AI think() cycle as long as delay is not greater then 2000. (the code comment below is wrong). here we would sort 3 times even if there is nothing to do. Performancewise this isn't good I believe. especially while taking into account we need to sort one time in advance of this code

Reply:
In fact, in old design and this new design more than 1 jobs were allowed. But here we set it explicitely to 3 or 5. More readable. Of course this is upper limit, we will run only those jobs that are overdue.

You are wrong - sort_task_pool() is run only once, but I changed design a bit, and in fact here was probably a problem before and should be fixed. Of course I will move it to proper position in next commit

Reply:
I found we need to run sort_task_pool twice, after all...

Reply:
ok I have to be more precise. in the old design sort_taskpool() was running only one sorting cycle for each call ensuring the oldest task in front. it was run one time before the block to evaluate delay time and then only if this was positive  a task was pushed to the temporary queue. Only after this push happened an next sort cycle was run to get the second oldest task in second place, just in case a second job had to be run. so we had a potentially useless sort cycle here as well.   
with your design we would sort 3 times or 5 times. then we would check the first 3 or 5 entries whether to push them to the queue just to realize there is nothing to do in worst case.  
I agree your decision algorithm is much more readable then the algorithm based on the sqrt of delay time but it is much simpler as well as it could only have 2 values 3 or 5. Furthermore this has the risk that if 3 jobs are due at the same time we will run all 3 of them blocking the logic functions for the complete time (at least as long as we can't run the AI ins its own thread. this will lead to the game seem to be halted. while if we only run 1 in each cycle the logic code could be run in between, this would mean the 2nd and third task might be delayed by about 500ms (if run at 1x speed) or multiples of this. so the current (old) design is better in terms of load distribution.       
So I believe the task to be added to the temporary queues were properly sorted (due to the call to sort_taskpool() after each push) if this was your main concern.

Reply:
> 

in this case just provide (jobs -1 ) to the second run

Reply:
This task pool is usually about 15 members, the sorting overhead is not that big. Based on my debugging, AI most of time has 3 jobs overdue, so unneeded sorting is very rare. But you are dead right about '-1', I will modify this.

I do not understand you concern about multiple jobs run within one AI think(). This is not new, it worked before...

Reply:
> 

as explained below the concern is not about this working. it is just about keeping the display smooth as long as possible.